src/utils/update/review.beneficial.owner.ts

```diff
@@ -0,0 +1,106 @@
+import { BeneficialOwnerOther, BeneficialOwnerOtherKey } from "../../model/beneficial.owner.other.model";
+import {
+  REVIEW_BENEFICIAL_OWNER_INDEX_PARAM,
+  UPDATE_AN_OVERSEAS_ENTITY_URL, UPDATE_REVIEW_BENEFICIAL_OWNER_INDIVIDUAL_PAGE, UPDATE_REVIEW_BENEFICIAL_OWNER_OTHER_PAGE,
+} from "../../config";
+import { ApplicationData } from "../../model";
+import { BeneficialOwnerIndividual, BeneficialOwnerIndividualKey } from "../../model/beneficial.owner.individual.model";
+
+const AllBoTypes = {
+  boiReview: "review_beneficial_owners_individual",
+  boIndividual: BeneficialOwnerIndividualKey,
+  booReview: "review_beneficial_owners_corporate",
+  boOther: BeneficialOwnerOtherKey
+};
+
+const beneficialOwnerIndividualReviewRedirectUrl = `${UPDATE_AN_OVERSEAS_ENTITY_URL
+      + UPDATE_REVIEW_BENEFICIAL_OWNER_INDIVIDUAL_PAGE
+      + REVIEW_BENEFICIAL_OWNER_INDEX_PARAM}`;
+
+const beneficialOwnerOtherReviewRedirectUrl = `${UPDATE_AN_OVERSEAS_ENTITY_URL
+        + UPDATE_REVIEW_BENEFICIAL_OWNER_OTHER_PAGE
+        + REVIEW_BENEFICIAL_OWNER_INDEX_PARAM}`;
+
+const checkBoIndividualValidation = (boi: BeneficialOwnerIndividual): boolean => {
+  return boi?.usual_residential_address ? true : false;
+};
+
+// this check is to determine whether the BO has been fully submitted
+// by checking for appData submitted with form and not present after PSC fetch
+const checkBoOtherValidation = (boOther: BeneficialOwnerOther): boolean => {
+  return boOther?.principal_address ? true : false;
+};
+
+const checkForBackButtonBo = (appData: ApplicationData, boType: string, boRedirectUrl: string) => {
+  const boLength: number = appData[boType]?.length || 0;
+  const boIndex = boLength - 1;
+  const isAppDataAndBoLength = (appData[boType] && boLength >= 1);
+
+  if (
+    isAppDataAndBoLength && (
+      ((boType === AllBoTypes.boIndividual) && !checkBoIndividualValidation(appData[boType][boIndex]))
+    ||
+      ((boType === AllBoTypes.boOther) && !checkBoOtherValidation(appData[boType][boIndex]))
+    )
+  ) {
+    return `${boRedirectUrl}${boIndex}`;
+  }
+};
+
+export const checkAndReviewBeneficialOwner = (appData: ApplicationData): string => {
+  let redirectUrl = "";
+
+  for (const updateAppData in appData.update){
```

Comment:
this is quite fragile as name changes could mess up the switch. Is it not the same as just going through the boi, boo then bo gov? I'd just use 3 code segments. Feel the for/swicth is redundant.

Reply:
to be addressed in Olayode's #869